### ISSUE TITLE
docker in docker build fails to mount

### ISSUE THREAD

**Issue by stevie-** [terraform-aws-modules/terraform-aws-lambda#42]
We want to use the terraform-aws-lambda inside docker in a Github Actions pipeline.

But this leads to an issue with mounts (check reference below). 

What we do is:

Github Actions Pipeline:
 runs a step with a docker container from github action. This container image includes all tools we require to run our terraform setup (including docker, aws cli and terragurnt etc).
 Inside this docker container terraform starts another docker container to build the lambda function with a custom requirements.txt.

We traced the error down to the docker mount for the temp_dir. 
The package.py copys the requirement.txt into a temp_dir https://github.com/terraform-aws-modules/terraform-aws-lambda/blob/47917cec498fb443166dbfb2fa3498de1d5e73ea/package.py#L848-L850

This temp.dir is passed on to the docker mount.
https://github.com/terraform-aws-modules/terraform-aws-lambda/blob/47917cec498fb443166dbfb2fa3498de1d5e73ea/package.py#L880-L885 

This works as long as long as the temp_dir matches the host dir. If we are inside a docker container the mount only works, if the path is same on docker host.

To make this work inside a docker container, we need to pass a path to the `build_root`, which exists on host too. For Example the `docker_build_root`. 

Request:
Is it possible to make the temp_dir path configurable for the `docker_run_command`?

---

Settings:
```
  function_name = "auth-rico-api-2"
  handler       = "app.lambda_handler"
  runtime       = "python3.8"
  source_path = [
    {
      path             = "${get_parent_terragrunt_dir()}/../auth-rico-api/src",
      pip_requirements = "${get_parent_terragrunt_dir()}/../auth-rico-api/requirements.txt",
    }
  ]

  build_in_docker   = true
  docker_image      = "lambci/lambda:build-python3.8"
  docker_build_root = "${get_parent_terragrunt_dir()}/../auth-rico-api/docker"
```

Github action log snippet
```
020-07-30T12:57:05.3081733Z [0m[1mmodule.lambda.data.external.archive_prepare[0]: Refreshing state...[0m
2020-07-30T12:57:05.4096716Z [0m[1mmodule.lambda.local_file.archive_plan[0]: Refreshing state... [id=2b948c89672589b46203225ec65cb807918153ed][0m
2020-07-30T12:57:05.4129139Z [0m[1mmodule.lambda.null_resource.archive[0]: Refreshing state... [id=7793903743368564043][0m
2020-07-30T12:57:06.2050799Z [0m[1mmodule.lambda.aws_cloudwatch_log_group.lambda[0]: Refreshing state... [id=/aws/lambda/auth-rico-api-2][0m
2020-07-30T12:57:06.2153055Z [0m[1mmodule.lambda.aws_lambda_function.this[0]: Refreshing state... [id=auth-rico-api-2][0m
2020-07-30T12:57:07.0441977Z [0m[1mmodule.lambda.aws_lambda_permission.unqualified_alias_triggers["APIGatewayAuth"]: Refreshing state... [id=APIGatewayAuth][0m
2020-07-30T12:57:09.6552211Z [0m[1mmodule.lambda.null_resource.archive[0]: Destroying... [id=7793903743368564043][0m[0m
2020-07-30T12:57:09.6608860Z [0m[1mmodule.lambda.null_resource.archive[0]: Destruction complete after 0s[0m[0m
2020-07-30T12:57:09.6686465Z [0m[1mmodule.lambda.null_resource.archive[0]: Creating...[0m[0m
2020-07-30T12:57:09.6717094Z [0m[1mmodule.lambda.null_resource.archive[0]: Provisioning with 'local-exec'...[0m[0m
2020-07-30T12:57:09.6737176Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): Executing: ["python3" ".terraform/modules/lambda/terraform-aws-lambda-1.17.0/package.py" "build" "--timestamp" "<WARNING: Missing lambda zip artifacts wouldn't be restored>" "builds/111221a0902b17abdaa47a73ee85cdd16429ada24bee754b20af73e6a048d174.plan.json"]
2020-07-30T12:57:09.7742867Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): CMD: python3 .terraform/modules/lambda/terraform-aws-lambda-1.17.0/package.py build --timestamp '<WARNING: Missing lambda zip artifacts wouldn'"'"'t be restored>' builds/111221a0902b17abdaa47a73ee85cdd16429ada24bee754b20af73e6a048d174.plan.json
2020-07-30T12:57:09.7752127Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): zip: creating 'builds/111221a0902b17abdaa47a73ee85cdd16429ada24bee754b20af73e6a048d174.zip' archive
2020-07-30T12:57:09.7759809Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): > docker images '--format={{.ID}}' lambci/lambda:build-python3.8
2020-07-30T12:57:09.8476378Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): > docker build --tag lambci/lambda:build-python3.8 /github/workspace/terraform/../auth-rico-api/docker
2020-07-30T12:57:09.8994543Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): Sending build context to Docker daemon   2.56kB
2020-07-30T12:57:09.9435179Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): Step 1/1 : FROM lambci/lambda:build-python3.8
2020-07-30T12:57:10.0856133Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): build-python3.8: Pulling from lambci/lambda
2020-07-30T12:57:10.0857150Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): b39539fa438b: Pulling fs layer
2020-07-30T12:57:10.0858368Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): 4c58555ee201: Pulling fs layer
2020-07-30T12:57:10.0859211Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): 7dec566ef15c: Pulling fs layer
2020-07-30T12:57:10.0860004Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): c9d5a7059e59: Pulling fs layer
2020-07-30T12:57:10.0860762Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): c9d5a7059e59: Waiting
2020-07-30T12:57:10.7658610Z [0m[1mmodule.lambda.aws_cloudwatch_log_group.lambda[0]: Modifying... [id=/aws/lambda/auth-rico-api-2][0m[0m
2020-07-30T12:57:10.9959614Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): b39539fa438b: Verifying Checksum
2020-07-30T12:57:10.9960830Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): b39539fa438b: Download complete
2020-07-30T12:57:10.9962178Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): 7dec566ef15c: Verifying Checksum
2020-07-30T12:57:10.9963154Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): 7dec566ef15c: Download complete
2020-07-30T12:57:11.9905031Z [0m[1mmodule.lambda.aws_cloudwatch_log_group.lambda[0]: Modifications complete after 1s [id=/aws/lambda/auth-rico-api-2][0m[0m
2020-07-30T12:57:12.1390788Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): c9d5a7059e59: Verifying Checksum
2020-07-30T12:57:12.1392221Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): c9d5a7059e59: Download complete
2020-07-30T12:57:14.7389790Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): 4c58555ee201: Verifying Checksum
2020-07-30T12:57:14.7390907Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): 4c58555ee201: Download complete
2020-07-30T12:57:17.6174603Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): b39539fa438b: Pull complete
2020-07-30T12:57:19.6703821Z [0m[1mmodule.lambda.null_resource.archive[0]: Still creating... [10s elapsed][0m[0m
2020-07-30T12:57:29.6704795Z [0m[1mmodule.lambda.null_resource.archive[0]: Still creating... [20s elapsed][0m[0m
2020-07-30T12:57:39.6704736Z [0m[1mmodule.lambda.null_resource.archive[0]: Still creating... [30s elapsed][0m[0m
2020-07-30T12:57:41.5042549Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): 4c58555ee201: Pull complete
2020-07-30T12:57:46.1174416Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): 7dec566ef15c: Pull complete
2020-07-30T12:57:49.6705380Z [0m[1mmodule.lambda.null_resource.archive[0]: Still creating... [40s elapsed][0m[0m
2020-07-30T12:57:53.7085098Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): c9d5a7059e59: Pull complete
2020-07-30T12:57:53.7122503Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): Digest: sha256:5a4cd0714a51e6d00059458a7f979ec5cbc96dfe1390e8488a1b6328ae3911cd
2020-07-30T12:57:53.7148080Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): Status: Downloaded newer image for lambci/lambda:build-python3.8
2020-07-30T12:57:53.7157454Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec):  ---> f1a7423d78df
2020-07-30T12:57:53.7161515Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): Successfully built f1a7423d78df
2020-07-30T12:57:53.7218594Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): Successfully tagged lambci/lambda:build-python3.8
2020-07-30T12:57:53.7220152Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): > docker images '--format={{.ID}}' lambci/lambda:build-python3.8
2020-07-30T12:57:53.7732536Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): DOCKER TAG ID: lambci/lambda:build-python3.8 -> f1a7423d78df
2020-07-30T12:57:53.7734535Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): Installing python requirements: /github/workspace/terraform/../auth-rico-api/requirements.txt
2020-07-30T12:57:53.7736133Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): > mktemp -d terraform-aws-lambda-XXXXXXXX # /tmp/terraform-aws-lambda-rqe8ucr7
2020-07-30T12:57:53.7737502Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): > cd /tmp/terraform-aws-lambda-rqe8ucr7
2020-07-30T12:57:53.7746448Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): > docker run --rm -v /tmp/terraform-aws-lambda-rqe8ucr7:/var/task:z -v /github/home/.ssh/known_hosts:/root/.ssh/known_hosts:z f1a7423d78df /bin/sh -c 'python3.8 -m pip install --no-compile --prefix= --target=. --requirement=requirements.txt && chown -R 0:0 .'
2020-07-30T12:57:59.6708092Z [0m[1mmodule.lambda.null_resource.archive[0]: Still creating... [50s elapsed][0m[0m
2020-07-30T12:58:00.9990088Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'requirements.txt'
2020-07-30T12:58:01.2113586Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): zip: Error during zip archive creation
2020-07-30T12:58:01.2134768Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): Traceback (most recent call last):
2020-07-30T12:58:01.2135948Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec):   File ".terraform/modules/lambda/terraform-aws-lambda-1.17.0/package.py", line 1116, in build_command
2020-07-30T12:58:01.2136975Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec):     bpm.execute(build_plan, zs, query)
2020-07-30T12:58:01.2138041Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec):   File ".terraform/modules/lambda/terraform-aws-lambda-1.17.0/package.py", line 764, in execute
2020-07-30T12:58:01.2139374Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec):     with install_pip_requirements(query, pip_requirements) as rd:
2020-07-30T12:58:01.2140439Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec):   File "/usr/lib/python3.8/contextlib.py", line 113, in __enter__
2020-07-30T12:58:01.2141383Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec):     return next(self.gen)
2020-07-30T12:58:01.2142652Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec):   File ".terraform/modules/lambda/terraform-aws-lambda-1.17.0/package.py", line 880, in install_pip_requirements
2020-07-30T12:58:01.2143637Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec):     check_call(docker_run_command(
2020-07-30T12:58:01.2168445Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec):   File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
2020-07-30T12:58:01.2172876Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec):     raise CalledProcessError(retcode, cmd)
2020-07-30T12:58:01.2179540Z [0m[0mmodule.lambda.null_resource.archive[0] (local-exec): subprocess.CalledProcessError: Command '['docker', 'run', '--rm', '-v', '/tmp/terraform-aws-lambda-rqe8ucr7:/var/task:z', '-v', '/github/home/.ssh/known_hosts:/root/.ssh/known_hosts:z', 'f1a7423d78df', '/bin/sh', '-c', 'python3.8 -m pip install --no-compile --prefix= --target=. --requirement=requirements.txt && chown -R 0:0 .']' returned non-zero exit status 1.
```

Reference:
https://serverfault.com/questions/819369/mounting-a-volume-with-docker-in-docker

**Comment by stevie-** [comment]
I wrote a `hack` which works for docker in docker on my machine and in github actions
https://github.com/terraform-aws-modules/terraform-aws-lambda/compare/master...datadrivers:feature/docker-in-docker-build

This may help to find a working solution for all cases

**Comment by ahlinc** [comment]
@stevie- thanks for the hack, it'll be useful on writing clear solution. I think there will be an additional environment variable(s) that will help to adapt package.py to run in a CI/CD environments.

Note for myself: https://rammusxu.github.io/toolkit/snippets/github-action/#2020-05-18-github-workflow-env-on-pull-request

**Comment by stevie-** [comment]
I've update the code to solve a bug and reuse temp_dir functionality. Goal is to create the temp_dir in the host path derived from requirements.txt path.

**Comment by sriramgk** [comment]
I am facing the same type of issue in the Gitlab pipeline. Is there any update on this?

**Comment by eperdeme** [comment]
Same issue here using Gitlab CI, shall have to use the fork/hack.

**Comment by eperdeme** [comment]
Resolved issue here without changing the source.

We use terragrunt and save the plan in 1 CI step and apply it in a 2nd step but it kept erroring saying couldn't find the tfplan file. We're now caching the entire .terragrunt folder and pasing it betwee plan+apply steps as this module writes our an additional tfplan file.

```yaml
      - ${CI_PROJECT_DIR}/**/*.tfplan # Needed for plan files from tf-aws-lambda-module

```

**Comment by stevie-** [comment]
> Resolved issue here without changing the source.
> 
> We use terragrunt and save the plan in 1 CI step and apply it in a 2nd step but it kept erroring saying couldn't find the tfplan file. We're now caching the entire .terragrunt folder and pasing it betwee plan+apply steps as this module writes our an additional tfplan file.
> 
> ```yaml
>       - ${CI_PROJECT_DIR}/**/*.tfplan # Needed for plan files from tf-aws-lambda-module
> ```

This is another issue not refering to the original problem in this issue ticket.

**Comment by Supercheez** [comment]
You can set a TMP_DIR environment variable to configure this, the underlying python package makes use of this and I was able to resolve Gitlab CI issues in this way.

**Comment by j2clerck** [comment]
Thank you @Supercheez for the tip on using TMP_DIR. For anyone else looking here, I found that using a relative path is incompatible with package.py which tries to delete the requirements file after switching directory [here](https://github.com/terraform-aws-modules/terraform-aws-lambda/blob/master/package.py#L966).
I also had issue with Terraform behavior when using an absolute path. Probably due to my build environment.
Finally I opted for using TMP_DIR=${env.WORKSPACE} in a Jenkins environment to make it work.
